File: puzzles/go/telephone-numbers.go
package main

import (
  "fmt"
  "strings"
  "strconv"
)

type Node struct {
  digit int
  children []Node
}

func index(node *Node, digit int) int {
  for i, child := range node.children {
    if child.digit == digit {
      return i
    }
  }
  return -1
}

func addChild(node *Node, digit int) {
  var children []Node
  var child = Node{digit, children}
  node.children = append(node.children, child)
}

func main() {
  var nbPhones int
  fmt.Scan(&nbPhones)
  var nbNodes = 0
  var children []Node
  var root = Node{-1, children}

  for i := 0; i < nbPhones; i++ {
    var telephone string
    fmt.Scan(&telephone)
    var slice = strings.Split(telephone, "")

    var current = &root
    for _, value := range slice {
      var digit, _ = strconv.Atoi(value)
      var i = index(current, digit)
      if i == -1 {
        addChild(current, digit)
        nbNodes++
        i = index(current, digit)
      }
      current = &(current.children[i])
    }
  }
  
  fmt.Println(nbNodes)
}

File: puzzles/js/telephone-numbers.js
class Node {
  constructor(digit) {
    this.digit = digit;
    this.children = [];
  }

  indexOf(digit) {
    for (let i = 0; i < this.children.length; ++i) {
      let child = this.children[i];
      if (child.digit === digit) {
        return i;
      }
    }
    return -1;
  }

  addChild(digit) {
    let child = new Node(digit);
    this.children.push(child);
  }
}

let size = parseInt(readline()); // The number of telephone numbers.
let elements = 0; // The number of elements (referencing a number) stored in the structure.
let root = new Node(-1); // root node of all the telephone numbers

for (let i = 0; i < size; ++i) {
  let telephone = readline();
  let current = root;
  for (let j = 0; j < telephone.length; ++j) {
    let digit = parseInt(telephone.charAt(j));
    let index = current.indexOf(digit);
    if (index === -1) {
      current.addChild(digit);
      ++elements;
    }
    index = current.indexOf(digit);
    current = current.children[index];
  }
}

print(elements);
